collagraph/component.py

```diff
@@ -89,6 +105,28 @@ def before_unmount(self):
     def render():  # pragma: no cover
         pass
 
+    def provide(self, values: dict = None, /, **kwargs):
+        if values:
+            self._provided.update(values)
+        self._provided.update(kwargs)
+
+    def inject(self, *args, **kwargs):
+        """Handle all kwargs as keys with default values."""
+        # The following assert can be disabled for production apps by
+        # enabling python optimization (e.g. `python -O ...`)
+        assert not any(arg in kwargs for arg in args), "Duplicate key injection"
+
+        for arg in [*args, *kwargs]:
+            assert not hasattr(self, arg), f"Attribute '{arg}' already exists"
+            provided = None
+            parent = self.parent
+            while parent is not None:
+                if arg in parent._provided:
+                    provided = parent._provided[arg]
+                    break
+                parent = parent.parent
+            setattr(self, arg, provided if provided is not None else kwargs.get(arg))
```

Comment:
Why is this setting an attribute, instead of returning the provided values?

Reply:
Convenience :)

Values that are provided are not 'reactive' (although the values that they contain might be), and Vue provides the injected values as attributes on `this`. Seems like a good idea to me.

Reply:
Well, only in the options API.

Reply:
Seams less magical, so less surprising, which is more better!

Reply:
I'm glad you agree :)

Reply:
Also results in way less code, which is even more better!